File: backend/src/main/java/com/polytest/core/TestGenerator.java
package com.polytest.core;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.time.Duration;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TestGenerator {

    public record GenerationResponse(String status, String sourceFile, String testFile, String framework, boolean cached, String error) {}

    private final String provider;
    private final String model;
    private final ConfigManager config;
    private final SyntaxValidator validator;
    private final String cacheDir = "cache";

    public TestGenerator(String providerOverride, String modelOverride) {
        this.config = new ConfigManager();
        this.config.loadConfig();
        this.provider = providerOverride != null ? providerOverride : config.get("llm.provider", "mock");
        this.model = modelOverride != null ? modelOverride : config.get("llm.model", "gemini-1.5-flash");
        this.validator = new SyntaxValidator();
        new File(cacheDir).mkdirs();
    }

    /**
     * Generates a unit test suite for the target source file, validates it, and writes it to disk.
     */
    public GenerationResponse generateTest(String sourceFilePath, String outputDir, String forcedFramework, boolean useCache) {
        File srcFile = new File(sourceFilePath);
        if (!srcFile.exists()) {
            return new GenerationResponse("failed", sourceFilePath, null, null, false, "Source file not found: " + sourceFilePath);
        }

        // 1. Detect language and target framework
        LanguageDetector detector = new LanguageDetector();
        LanguageDetector.LanguageInfo langInfo = detector.detect(sourceFilePath);
        
        if (langInfo.language().equals("Unsupported")) {
            return new GenerationResponse("failed", sourceFilePath, null, null, false, "Extension or structure not supported: " + sourceFilePath);
        }

        String framework = forcedFramework != null ? forcedFramework : langInfo.framework();

        // 2. Parse source structure
        SourceParser parser = new SourceParser();
        SourceParser.ParseResult parseRes = parser.parse(sourceFilePath, langInfo.language());

        // 3. Assemble LLM Prompt
        String prompt = buildPrompt(srcFile.getName(), langInfo.language(), framework, parseRes);
        String promptHash = getMd5Hash(prompt);
        Path cachePath = Paths.get(cacheDir, promptHash + ".log");

        // 4. Query caching layer
        if (useCache && Files.exists(cachePath)) {
            try {
                String cachedCode = Files.readString(cachePath);
                String testDest = writeTestFile(srcFile.getName(), cachedCode, outputDir, langInfo.language());
                return new GenerationResponse("success", sourceFilePath, testDest, framework, true, null);
            } catch (IOException e) {
                // Fail-over to API query
            }
        }

        // 5. Query LLM provider
        String generatedCode;
        try {
            if (provider.equalsIgnoreCase("mock")) {
                generatedCode = generateMockStub(srcFile.getName(), langInfo.language(), framework, parseRes);
            } else {
                generatedCode = queryLLMApi(prompt);
            }
        } catch (Exception e) {
            return new GenerationResponse("failed", sourceFilePath, null, framework, false, "LLM query error: " + e.getMessage());
        }

        // Clean generated markdown decorators
        generatedCode = cleanMarkdownDecorators(generatedCode);

        // 6. Run syntax linter check
        SyntaxValidator.ValidationResult valRes = validator.validate(generatedCode, langInfo.language());
        if (!valRes.isValid()) {
            return new GenerationResponse("failed", sourceFilePath, null, framework, false, "Linter validation failed: " + valRes.errorLog());
        }

        // 7. Write to cache
        try {
            Files.writeString(cachePath, generatedCode);
        } catch (IOException e) {
            // Ignore cache write errors
        }

        // 8. Write to target destination folder
        try {
            String testDest = writeTestFile(srcFile.getName(), generatedCode, outputDir, langInfo.language());
            return new GenerationResponse("success", sourceFilePath, testDest, framework, false, null);
        } catch (IOException e) {
            return new GenerationResponse("failed", sourceFilePath, null, framework, false, "Failed to write target test file: " + e.getMessage());
        }
    }

    private String buildPrompt(String fileName, String language, String framework, SourceParser.ParseResult parseRes) {
        StringBuilder sb = new StringBuilder();
        sb.append("You are PolyTest AI, an advanced multi-language test generator.\n");
        sb.append("Generate a highly comprehensive, fully functional unit test suite in ").append(language);
        sb.append(" using the framework ").append(framework).append(" for the file: '").append(fileName).append("'.\n\n");
        
        sb.append("Analyzed code structure:\n");
        for (SourceParser.ClassInfo cls : parseRes.classes()) {
            sb.append("- Class: ").append(cls.name()).append(" with methods: ").append(cls.methods()).append("\n");
        }
        if (!parseRes.functions().isEmpty()) {
            sb.append("- Standalone functions: ").append(parseRes.functions()).append("\n");
        }
        sb.append("- Imports: ").append(parseRes.imports()).append("\n\n");
        
        sb.append("Ensure you import the core class/functions perfectly. Return ONLY the written valid code without markdown wrapping blocks.");
        return sb.toString();
    }

    private String generateMockStub(String fileName, String language, String framework, SourceParser.ParseResult parseRes) {
        String baseName = fileName.substring(0, fileName.lastIndexOf('.'));
        String className = parseRes.classes().isEmpty() ? "SourceModule" : parseRes.classes().get(0).name();
        
        switch (language.toLowerCase()) {
            case "python":
                return String.format("""
                    # Generated automatically by PolyTest AI mock engine for %s
                    import pytest
                    from %s import %s
                    
                    def test_initial_sanity():
                        # Basic assertion checks
                        assert True
                    
                    def test_dynamic_mock_cases():
                        # Mocking methods logic checks
                        assert 1 == 1
                    """, fileName, baseName, className);
            case "javascript":
            case "typescript":
                return String.format("""
                    // Generated automatically by PolyTest AI mock engine for %s
                    const { %s } = require('./%s');
                    
                    describe('Sanity Mock Test Suite', () => {
                        test('should verify basic assertions', () => {
                            expect(true).toBe(true);
                        });
                    });
                    """, fileName, className, baseName);
            case "go":
                return String.format("""
                    package main
                    // Generated automatically by PolyTest AI mock engine for %s
                    import "testing"
                    
                    func TestSanityCheck(t *testing.T) {
                        if false {
                            t.Errorf("Error during validation assertion")
                        }
                    }
                    """, fileName);
            case "java":
                return String.format("""
                    package tests;
                    // Generated automatically by PolyTest AI mock engine for %s
                    import org.junit.jupiter.api.Test;
                    import static org.junit.jupiter.api.Assertions.*;
                    
                    public class %sTest {
                        @Test
                        public void testSanity() {
                            assertTrue(true);
                        }
                    }
                    """, fileName, className);
            case "c++":
                return String.format("""
                    // Generated automatically by PolyTest AI mock engine for %s
                    #include <gtest/gtest.h>
                    #include "%s"
                    
                    TEST(MockSanitySuite, TestTrue) {
                        EXPECT_TRUE(true);
                    }
                    """, fileName, fileName);
            default:
                return "// Generated mock for: " + fileName;
        }
    }

    private String queryLLMApi(String prompt) throws Exception {
        // Native Java 11 HTTP client queries
        String apiKey = System.getenv("GEMINI_API_KEY");
        if (apiKey == null) apiKey = config.get("llm.api_key", "");
        
        if (apiKey.isEmpty()) {
            throw new IllegalStateException("API key not configured in environment or polytest.yaml for provider " + provider);
        }

        String endpoint = "";
        String jsonPayload = "";
        
        if (provider.equalsIgnoreCase("gemini")) {
            endpoint = "https://generativelanguage.googleapis.com/v1beta/models/" + model + ":generateContent?key=" + apiKey;
            jsonPayload = String.format("{\"contents\": [{\"parts\": [{\"text\": \"%s\"}]}]}", escapeJson(prompt));
        } else if (provider.equalsIgnoreCase("openai")) {
            endpoint = "https://api.openai.com/v1/chat/completions";
            jsonPayload = String.format("{\"model\": \"%s\", \"messages\": [{\"role\": \"user\", \"content\": \"%s\"}]}", model, escapeJson(prompt));
        } else {
            throw new IllegalArgumentException("Provider integrations not supported in java port: " + provider);
        }

        HttpClient client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonPayload))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new IOException("HTTP API response code " + response.statusCode() + ":\n" + response.body());
        }

        return parseLLMResponse(response.body());
    }

    private String parseLLMResponse(String body) {
        // Safe, native parsing without heavy JSON parsers
        if (provider.equalsIgnoreCase("gemini")) {
            Matcher matcher = Pattern.compile("\"text\"\\s*:\\s*\"(.*?)\"").matcher(body);
            if (matcher.find()) {
                return unescapeJson(matcher.group(1));
            }
        } else if (provider.equalsIgnoreCase("openai")) {
            Matcher matcher = Pattern.compile("\"content\"\\s*:\\s*\"(.*?)\"").matcher(body);
            if (matcher.find()) {
                return unescapeJson(matcher.group(1));
            }
        }
        return body;
    }

    private String writeTestFile(String srcFileName, String code, String outputDir, String language) throws IOException {
        new File(outputDir).mkdirs();
        
        String baseName = srcFileName.substring(0, srcFileName.lastIndexOf('.'));
        String ext = getExtension(language);
        String testFileName;

        if (language.equalsIgnoreCase("python")) {
            testFileName = "test_" + baseName + ext;
        } else if (language.equalsIgnoreCase("javascript") || language.equalsIgnoreCase("typescript")) {
            testFileName = baseName + ".test" + ext;
        } else if (language.equalsIgnoreCase("go")) {
            testFileName = baseName + "_test" + ext;
        } else {
            testFileName = baseName + "Test" + ext;
        }

        Path dest = Paths.get(outputDir, testFileName);
        Files.writeString(dest, code);
        return dest.toAbsolutePath().toString();
    }

    private String cleanMarkdownDecorators(String code) {
        return code.replaceAll("(?s)```[a-zA-Z]*\\s*", "").replaceAll("```", "").trim();
    }

    private String escapeJson(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private String unescapeJson(String s) {
        return s.replace("\\n", "\n")
                .replace("\\r", "\r")
                .replace("\\\"", "\"")
                .replace("\\\\", "\\");
    }

    private String getMd5Hash(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hashBytes = md.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : hashBytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (Exception e) {
            return String.valueOf(input.hashCode());
        }
    }

    private String getExtension(String language) {
        switch (language.toLowerCase()) {
            case "python": return ".py";
            case "javascript": return ".js";
            case "typescript": return ".ts";
            case "go": return ".go";
            case "java": return ".java";
            case "c++": return ".cpp";
            case "c#": return ".cs";
            default: return ".txt";
        }
    }
}

File: core/generator/test_generator.py
import os
import shutil
from typing import Dict, Any, Optional
from core.analyzer.detector import LanguageDetector, LanguageInfo
from core.analyzer.parser_factory import ParserFactory
from core.context.context_builder import ContextBuilder
from core.llm.llm_client import LLMClient
from core.llm.prompt_builder import PromptBuilder
from core.llm.response_parser import ResponseParser

class TestGenerator:
    """Orchestrates the entire language-detection, code-parsing, prompt-building, and test-generation pipeline."""

    def __init__(self, 
                 provider: str = "mock", 
                 model: Optional[str] = None, 
                 endpoint: Optional[str] = None,
                 cache_dir: Optional[str] = "cache"):
        self.provider_name = provider
        self.model_name = model
        self.endpoint = endpoint
        self.cache_dir = cache_dir
        
        self.detector = LanguageDetector()
        self.context_builder = ContextBuilder()
        self.prompt_builder = PromptBuilder()
        self.response_parser = ResponseParser()
        
        if self.cache_dir:
            os.makedirs(self.cache_dir, exist_ok=True)

    def generate_test(self, 
                      source_file_path: str, 
                      output_dir: str, 
                      forced_framework: Optional[str] = None, 
                      use_cache: bool = True) -> Dict[str, Any]:
        """Generate a complete test suite for a single source code file.
        
        Args:
            source_file_path: Absolute or relative path to the target source file.
            output_dir: Directory where the generated test file should be saved.
            forced_framework: Force a testing framework instead of auto-detecting it.
            use_cache: Whether to cache LLM responses locally to save cost.
            
        Returns:
            A results dictionary with status, file path, language, framework, and metrics.
        """
        if not os.path.isfile(source_file_path):
            return {
                "status": "failed",
                "error": f"Source file does not exist: {source_file_path}"
            }

        # 1. Detect language & framework
        lang_info = self.detector.detect(source_file_path)
        target_framework = forced_framework or lang_info.framework
        
        # 2. Read source file
        try:
            with open(source_file_path, 'r', encoding='utf-8') as f:
                source_content = f.read()
        except Exception as e:
            return {
                "status": "failed",
                "error": f"Failed to read source file: {e}"
            }

        # 3. Parse file structure
        parser = ParserFactory.get_parser(lang_info.language)
        parser_result = parser.parse(source_content, source_file_path)

        # 4. Build rich context
        context = self.context_builder.build_context(
            file_path=source_file_path,
            file_content=source_content,
            parser_result=parser_result,
            target_framework=target_framework
        )

        # 5. Determine cached log path
        file_name = os.path.basename(source_file_path)
        cache_log_path = None
        if self.cache_dir:
            cache_log_path = os.path.join(self.cache_dir, f"{file_name}.log")

        llm_response = None
        loaded_from_cache = False

        # 6. Check cache first
        if use_cache and cache_log_path and os.path.exists(cache_log_path):
            try:
                with open(cache_log_path, 'r', encoding='utf-8') as cf:
                    llm_response = cf.read()
                loaded_from_cache = True
            except Exception:
                pass

        # 7. Query LLM if not cached
        if not llm_response:
            # Build prompt
            system_instruction = self.prompt_builder.build_system_instruction(context)
            user_prompt = self.prompt_builder.build_user_prompt(context)
            
            # Fetch client provider
            provider_client = LLMClient.get_provider(
                provider_name=self.provider_name,
                model=self.model_name,
                endpoint=self.endpoint
            )
            
            # Generate
            llm_response = provider_client.generate(user_prompt, system_instruction=system_instruction)
            
            # Save cache
            if use_cache and cache_log_path and llm_response and "Error" not in llm_response:
                try:
                    with open(cache_log_path, 'w', encoding='utf-8') as cf:
                        cf.write(llm_response)
                except Exception:
                    pass

        if not llm_response:
            return {
                "status": "failed",
                "error": "Empty or failed response from the LLM Provider."
            }

        # 8. Parse the code response
        generated_test_code = self.response_parser.extract_code(llm_response)

        # 9. Format output test file name
        test_file_name = self._format_test_filename(file_name, lang_info.language, target_framework)
        os.makedirs(output_dir, exist_ok=True)
        test_file_path = os.path.join(output_dir, test_file_name)

        # 10. Write the test suite
        try:
            with open(test_file_path, 'w', encoding='utf-8') as tf:
                tf.write(generated_test_code)
        except Exception as e:
            return {
                "status": "failed",
                "error": f"Failed to write generated test file: {e}"
            }

        return {
            "status": "success",
            "source_file": source_file_path,
            "test_file": test_file_path,
            "language": lang_info.language,
            "framework": target_framework,
            "complexity": parser_result.estimated_complexity,
            "lines_of_code": parser_result.lines_of_code,
            "classes_found": len(parser_result.classes),
            "functions_found": len(parser_result.functions),
            "cached": loaded_from_cache
        }

    def _format_test_filename(self, source_filename: str, language: str, framework: str) -> str:
        """Format the output test file name according to language and framework standards."""
        name_no_ext, ext = os.path.splitext(source_filename)
        lang_lower = language.lower()
        
        if lang_lower == 'python':
            # test_*.py
            if source_filename.startswith('test_'):
                return source_filename
            return f"test_{name_no_ext}{ext}"
            
        elif lang_lower in ('javascript', 'typescript'):
            # *.test.js or *.test.ts
            if '.test.' in source_filename or '.spec.' in source_filename:
                return source_filename
            return f"{name_no_ext}.test{ext}"
            
        elif lang_lower == 'c++':
            # *_test.cpp or test_*.cpp
            if '_test' in name_no_ext or 'test_' in name_no_ext:
                return source_filename
            return f"{name_no_ext}_test{ext}"
            
        elif lang_lower == 'java':
            # *Test.java
            if name_no_ext.endswith('Test'):
                return source_filename
            return f"{name_no_ext}Test{ext}"
            
        elif lang_lower == 'go':
            # *_test.go
            if name_no_ext.endswith('_test'):
                return source_filename
            return f"{name_no_ext}_test{ext}"
            
        elif lang_lower == 'c#':
            # *Tests.cs or *Test.cs
            if name_no_ext.endswith('Tests') or name_no_ext.endswith('Test'):
                return source_filename
            return f"{name_no_ext}Tests{ext}"
            
        else:
            return f"test_{source_filename}"
